src/main/java/org/skriptlang/skript/scheduler/platforms/SpigotScheduler.java

```diff
@@ -0,0 +1,158 @@
+/**
+ *   This file is part of Skript.
+ *
+ *  Skript is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  Skript is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Skript.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright Peter Güttinger, SkriptLang team and contributors
+ */
+package org.skriptlang.skript.scheduler.platforms;
+
+import java.util.WeakHashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+import org.bukkit.Bukkit;
+import org.bukkit.plugin.Plugin;
+import org.eclipse.jdt.annotation.Nullable;
+import org.skriptlang.skript.scheduler.AsyncTask;
+import org.skriptlang.skript.scheduler.PlatformScheduler;
+import org.skriptlang.skript.scheduler.Task;
+
+import ch.njol.skript.Skript;
+
+public class SpigotScheduler implements PlatformScheduler {
+
+	private static final Map<Integer, AsyncTask> asyncTasks = new WeakHashMap<>();
```

Comment:
This is not what weak hashmaps do. The (wrapped integer) keys are weak references, the values are strong references.
You do not want weak integer references since the memory behaviour of these constables will be shaky at best. In fact, it's very likely that these could never be garbage-collected even in a perfect world since the hash table calculation will probably end up referencing the same Integer object.

If you want the values to be weak you need a regular (hash) map of integers to References, and then wrap each *value* in a weak reference. Of course, this will change the entry set behaviour so you might want to create your own map wrapper with a special kind of entry set. I made a rudimentary example of this [here](https://github.com/Moderocky/Clockwork/blob/master/src/main/java/mx/kenzie/clockwork/collection/ReferenceMap.java#L9) a while ago.

However, remember that a map of weak or soft references will leave a lot of stubs (e.g. the key, the reference object) that are *not* garbage collected, so to avoid the map filling up with garbage you may want some kind of occasional emptying method on a timer.